### DISCUSSION TITLE
import.meta.url and Webpack 5: avoiding absolute paths

### DISCUSSION THREAD

**Discussion by PhilippWendler** [gildas-lormeau/zip.js#299] [General]
I am using this great library for some time and recently have started using Webpack 5 (via create-react-app). I am interested only in unzipping and I do not want web workers, so I use the following in my code:
```js
import * as zip from "@zip.js/zip.js/lib/zip-no-worker-inflate";
```
Everything works fine.

However: When I bundle my application with Webpack, Webpack inserts the absolute path to `codec-pool-worker.js` into the generated code, because that file uses `import.meta.url`:
https://github.com/gildas-lormeau/zip.js/blob/f83228cf0029cd2f0e47bb0a9419f31666e47f8b/lib/core/codecs/codec-pool-worker.js#L112
This is highly undesired for me, because it leaks my user name and directory names and it leads to the build being different depending on where I build it, for example on my machine vs. in CI (everything else in the build is deterministic and [reproducible](https://reproducible-builds.org/)). From what I understand from the Webpack issue tracker this is by design and not changeable.

I can work around this if I copy `codec-pool-worker.js` to my app, delete `createWebWorkerInterface` and tell Webpack to use my copy of the file instead of the original one. However, this is fragile and I of course would prefer to avoid this.

Given that the code in `createWebWorkerInterface` should not be used anyway when using any of `lib/zip-no-worker*` AFAIU, I wondered whether there is a possibility for the library to be restructured in a way such that `createWebWorkerInterface` can be removed by tree shaking when `lib/zip-no-worker*` is imported? This would also reduce the bundle size by a few bytes, of course. This is why I am raising this here.

Alternatively, if `createWebWorkerInterface` would be in a separate file this would be easy for me, though I could understand if you don't want to make such adjustments for purely external reasons.

Or does anybody have an idea what else I could do to avoid ending up with the absolute path in my bundle, without having to copy and change the code? I am not really an expert in this.

**Comment by gildas-lormeau** [comment]
> From what I understand from the Webpack issue tracker this is by design and not changeable.

If I'm not mistaken, the Webpack team recognizes that this is a bug, see https://github.com/webpack/webpack/issues/14445. For the record, I use Rollup to build zip.js and I don't have this problem.
To circumvent this issue, maybe you could replace `new URL(workerData.scripts[0], import.meta.url)` with `workerData.scripts[0]` from the source code with this plugin https://www.npmjs.com/package/webpack-plugin-replace or something similar?

**Reply by gildas-lormeau** [reply]
FYI, I'm fixing the tree-shaking issue.

**Comment by gildas-lormeau** [comment]
It's fixed in the version 2.4.0.

**Reply by PhilippWendler** [reply]
Very nice! I just upgraded to 2.4.1 and everything is as it should be. Thank you very much!

**Reply by gildas-lormeau** [reply]
You're welcome :)

**Reply by Chetanr** [reply]
Hi,
I am using webpack 2.6.35 and the issue still persists.
<img width="1433" alt="Screen Shot 2022-10-06 at 10 43 54 am" src="https://user-images.githubusercontent.com/5353603/194183292-cb6b1e0e-cca5-418f-b9db-bc0d952bfd25.png">

**Reply by gildas-lormeau** [reply]
Create a file named `zip.js` in your project containing the following content:

```js
import { configure } from "@zip.js/zip.js/lib/core/configuration.js";

// remove the 2 following lines if you don't use web workers
import { configureWebWorker } from "@zip.js/zip.js/lib/z-worker-inline.js";
import { terminateWorkers } from "@zip.js/zip.js/lib/core/codec-pool.js";

// uncomment the 2 following lines if you don't use web workers
// import { Deflate } from "@zip.js/zip.js/lib/core/streams/codecs/deflate.js";
// import { Inflate } from "@zip.js/zip.js/lib/core/streams/codecs/inflate.js";

// remove the following line if you don't use web workers
configureWebWorker(configure);

// uncomment the following line if you don't use web workers
// configure({ Deflate, Inflate });

export * from "@zip.js/zip.js/lib/core/io.js";
export * from "@zip.js/zip.js/lib/core/zip-reader.js";
export * from "@zip.js/zip.js/lib/core/zip-writer.js";
export {
  configure,
  // remove the following line if you don't use web workers
  terminateWorkers
};
```

Then, replace the import(s) of zip.js in your project with this file.

For example, replace:
```js
import * as zip from "@zip.js/zip.js";
```
with:
```js
import * as zip from "./zip.js";
```

**Reply by Chetanr** [reply]
I am only using BlobReader, TextWriter and ZipReader.

how do I do it in that case?

**Reply by gildas-lormeau** [reply]
If your build tool supports tree shaking, you don't have to care about this. You just have to write the code below.

```js
import { BlobReader, TextWriter, ZipReader } from "./zip.js";
```

**Reply by Chetanr** [reply]
Tried that but it
<img width="1433" alt="Screen Shot 2022-10-06 at 11 13 59 am" src="https://user-images.githubusercontent.com/5353603/194186061-38906ccd-70cc-48fb-8292-8b30c810b7fd.png">

Here is zip.ts file
<img width="1125" alt="Screen Shot 2022-10-06 at 11 14 34 am" src="https://user-images.githubusercontent.com/5353603/194186115-dd726d2c-6a7a-412c-9bba-65aaefc33c80.png">

**Reply by gildas-lormeau** [reply]
I don't know why it does not work for you. Here is a simple project that works for me: [working-example.zip](https://github.com/gildas-lormeau/zip.js/files/9725929/working-example.zip)
